### ISSUE TITLE
StackOverflowException when running tests with nunit3-console.exe

### ISSUE THREAD

**Issue by kdubau** [nunit/nunit#1045]
My apologies in advance if this I'm missing any information here. Please let me know what I can provide to help get to the bottom of this. I am trying to use the nunin3-console.exe runner, but am receiving a stack overflow exception immediately. Interestingly, this works fine on my ElCapitan iMac.

```console
> ../packages/NUnit.Console.3.0.0/tools/nunit3-console.exe" "bin/Debug/XQA.XS.Tests.dll" -where="cat == DotNet" -verbose
NUnit Console Runner 3.0.5797
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.10240.0
  CLR Version: 4.0.30319.42000

Test Files
    bin/Debug/XQA.XS.Tests.dll

Test Filters
    Where: cat == DotNet

Process is terminated due to StackOverflowException.

Errors and Failures

1) Error :
An existing connection was forcibly closed by the remote host

Server stack trace:
   at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)
   at System.Runtime.Remoting.Channels.SocketStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadFromSocket(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadAndMatchFourBytes(Byte[] buffer)
   at System.Runtime.Remoting.Channels.Tcp.TcpSocketHandler.ReadAndMatchPreamble()
   at System.Runtime.Remoting.Channels.Tcp.TcpSocketHandler.ReadVersionAndOperation(UInt16& operation)
   at System.Runtime.Remoting.Channels.Tcp.TcpClientSocketHandler.ReadHeaders()
   at System.Runtime.Remoting.Channels.Tcp.TcpClientTransportSink.ProcessMessage(IMessage msg, ITransportHeaders requestHeaders, Stream requestStream, ITransportHeaders& responseHeaders, Stream& responseStream)
   at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
    Overall result: Failed
   Tests run: 0, Passed: 0, Errors: 0, Failures: 0, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2015-11-20 01:05:37Z
    End time: 2015-11-20 01:05:45Z
    Duration: 8.659 seconds

Results (nunit3) saved as TestResult.xml
```

**Comment by kdubau** [comment]
My test is an automated UI test which remotes to the application under test. I was able to do some more debugging and just when my test starts the remote is when the exception is thrown.

**Comment by kdubau** [comment]
Additionally, after working around https://github.com/nunit/nunit3-vs-adapter/issues/78 the remoting session starts normally from the VS Test Explorer.

**Comment by CharliePoole** [comment]
What happens if you add --inprocess to your command-line?

**Comment by kdubau** [comment]
Hi Charlie, 

Seems I get the same problem but it does not print the stacktrace and sumamry. 
```
>"../packages/NUnit.Console.3.0.0/tools/nunit3-console.exe" "bin/Debug/XQA.XS.Tests.dll" --inprocess -where="cat == DotNet"
NUnit Console Runner 3.0.5797
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.10240.0
  CLR Version: 4.0.30319.42000

Test Files
    bin/Debug/XQA.XS.Tests.dll

Test Filters
    Where: cat == DotNet


Process is terminated due to StackOverflowException.
```

**Comment by CharliePoole** [comment]
That's good. We can't show any more info about it because we can't really take back control of the process after a stack overflow.

Next, add --labels=On so you can see whether any tests execute at all.

**Comment by kdubau** [comment]
Hmmm. Same thing, this is interesting because I thought my test was at least _starting_ and blowing up when I start my remote session. If it helps at all, my remote session starts in my `[SetUp]` method.
```
> "../packages/NUnit.Console.3.0.0/tools/nunit3-console.exe" "bin/Debug/XQA.XS.Tests.dll" --inprocess --labels=On -where="cat == DotNet"
NUnit Console Runner 3.0.5797
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.10240.0
  CLR Version: 4.0.30319.42000

Test Files
    bin/Debug/XQA.XS.Tests.dll

Test Filters
    Where: cat == DotNet


Process is terminated due to StackOverflowException.
```

**Comment by CharliePoole** [comment]
OK, drop the --labels and let's try --explore. That way NUnit won't even **try** to run tests. Any stack-overflow there has to be either in NUnit itself or in some load-time code of yours like TestCaseSource. Here's hoping it's you. :-)

**Comment by LordJZ** [comment]
@kdubau Could you please share how you worked around this issue? https://github.com/nunit/nunit3-vs-adapter/issues/78

**Comment by kdubau** [comment]
Yep. That seems to work correctly. It prints a list of every test in the assembly.

To recap, my test is an automated UI test and uses .NET remoting to remote to the application under test. From my debugging, I determined the StackOverflowException is thrown right when I make my first call to the remote object. If my understanding is correct, the nunit3-console.exe app uses remoting itself to the nunit engine - is this "remoting within remoting" causing the issue?

**Comment by kdubau** [comment]
@LordJZ I temporarily removed the `[TestSourceAttribute]` replacing it with standard techniques just to see if my tests ran at all from the adapter.

**Comment by kdubau** [comment]
@CharliePoole Thanks a lot for the support on this issue. Turns out my Nunit project was comping to AnyCpu and therefore affected by this [0] and this rare case caused the application to fall into an infinite loop during startup. It worked on mac obviously because there is no registry involved. And finally I have no idea why this didn't happen with Nunit 2.6 :)

[0] http://stackoverflow.com/questions/2464358/why-is-opensubkey-returning-null-on-my-windows-7-64-bit-system

**Comment by rprouse** [comment]
Could you try running your tests with the `--inprocess` flag, it might give us a more useful stack trace.

**Comment by Wyulliam** [comment]
Good morning.
I'm facing the same (or almost the same) issue descripted by the ones above.

I have a lot of tests to run, and, after running exactly 59 tests, the test number 60 gives me a Stack Overflow Exception, causing the program to crash, and the CMD does not stop runing (he runned for a full weekend after the error).
The problem is that if I run the test number 60 alone, of if a few more tests, he works with no exception. 
Any leads of what could that be?

(here is the exemple of what I got. I runned a propsital wrong test, causing a proposital Stack Overflow Exception)

NUnit Console Runner 3.2.1
Copyright (C) 2016 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.10240.0
  CLR Version: 4.0.30319.42000

Test Files
    AMBEV.AS.Tests.dll

Starting ChromeDriver (v2.9.248315) on port 63822

Process is terminated due to StackOverflowException.

**Comment by CharliePoole** [comment]
@Wyulliam if you throw a StackOvrrflowException then the process will terminate. That's unavoidable. The issue is about NUnit itself possibly throwing such an exception. It's closed because we couldn't find evidence that was happening.

**Comment by chadmyers** [comment]
I'd like to report that we're running into this as well. It seems sporadic.

We use jenkins-ec2 to spin up AWS EC2 instances for testing so we get fresh boxes all the time. It seems that once one of the machines has this issue, it'll keep having it.   But there are times (like today) where it happens consistently on all of our build slave VMs even if I terminate them and spin up new ones.  The next day it'll be fine and we won't have any problems then it comes in clusters again.

It seems worse in the afternoon/evening if that makes any sense. Maybe it's a time thing. All our VM's are using UTC as the time zone if that matters at all.  None of that makes any sense but those are the facts in case anyone else happens to notice similar trends.

**Comment by chadmyers** [comment]
It definitely seems to be an NUnit3 thing, though. We have similar code running in NUnit2.x and we don't have any problems like this. This is a new project sharing some testing code but using NUnit3 and then it blows up.

**Comment by rprouse** [comment]
We are getting enough reports about this that I think it warrants investigation. We may need a reproducible test case, or someone with a call stack otherwise we are blind on this one.

I am reopening with a confirm status. If anyone has ideas on how to pin this down or more information, please add to this issue.

**Comment by rprouse** [comment]
@chadmyers can you give us more information on your tests? Are you using `--where`? Are you testing a web app with a web framework like Selenium?

**Comment by chadmyers** [comment]
@rprouse 

Here's some more facts as I think of them (kind of a stream / thinking out loud):

 - it only seems to happen on our test slaves (Win2008 R2) and even then, not consistently
 - when we run the same tests and commands on our local dev boxes, we never see it
 - we are using Selenium WebDriver, yes.
 - We can't get a proper stack trace because it's a stack overflow, unfortunately. What we see is just like what the others posted above.
 - When I RDP into a test slave box and run the tests from a console:
   - I will sometimes see the stackoverflow, but not always.
   - I will sometimes see nunit-agent.exe crash and I get the windows pop up about a program crashing
   - If nunit-agent.exe hasn't crashed, and I kill the process, then nunit3-console.exe will stop hanging and then dump a stack trace and then exit. The stack trace is all having to do with remoting. I don't have the stack trace handy. We're going to be trying some more things today to try to pin this down and so I should have more information later to share with you

**Comment by akarnokd** [comment]
Not sure if related, but I have 91 test in a project but the test runner [fails](https://travis-ci.org/reactor/reactor-core-dotnet/builds/138090396); It worked [yesterday](https://travis-ci.org/reactor/reactor-core-dotnet/builds/137929007) with just 75 tests. On my machine running Windows 7 x64, Visual Studio 2015 and the latest NUnit3 runner plugin, everything is fine.

**Comment by ckimball** [comment]
Unsure if this is related, but I am seeing StackOverflowException being thrown from nunit console runner v3.2.1 whenever using the .nunit configuration file to specify assemblies. It works fine whenever I manually specify the assemblies.

**Comment by rprouse** [comment]
@ckimball is that in conjunction with any other command line options? Mind sharing your command line?

**Comment by chadmyers** [comment]
I saw this in the Event Log:

> nunit-agent.exe - System Error 
>  A new guard page for the stack cannot be created.  

There's not much else in there (no stack trace, etc).

Also, good news, one of our developers was able to get it to do stackoverflowexception locally on his dev machine so we should hopefully be able to get a more consistent failure and work towards narrowing down exactly what the trigger is

**Comment by chadmyers** [comment]
Here are the log files from a failed run:

[InternalTrace.1916.txt](https://github.com/nunit/nunit/files/319111/InternalTrace.1916.txt)
[InternalTrace.2436.esp.UXTesting.dll.txt](https://github.com/nunit/nunit/files/319112/InternalTrace.2436.esp.UXTesting.dll.txt)
[nunit-agent_2436.txt](https://github.com/nunit/nunit/files/319113/nunit-agent_2436.txt)

One thing I noticed was that it crashed just after a test completed and it was starting a new test.  Our log output doesn't show logging output from the teardown of the finished test or the setup of the new test. We just get the last few log entries from the finished test.

**Comment by chadmyers** [comment]
Here's our commandline:

```
nunit3-console .\tests\esp.UXTesting\bin\release\esp.UXTesting.dll --result:%WORKSPACE%\TestResult.xml;format=nunit2 /err:StdErr.txt /trace=DEBUG
```

**Comment by chadmyers** [comment]
I ran that command-line while RPD'd interactively into the testing slave and part-way through it I saw the "NUnit Agent has stopped working" pop up. I clicked "Show problem details" and it had this:

```
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	nunit-agent.exe
  Application Version:	3.2.0.0
  Application Timestamp:	56db3093
  Fault Module Name:	clr.dll
  Fault Module Version:	4.6.1076.0
  Fault Module Timestamp:	56d79ed4
  Exception Code:	c00000fd
  Exception Offset:	000000000000b26a
  OS Version:	6.1.7601.2.1.0.400.8
  Locale ID:	1033
  Additional Information 1:	cd18
  Additional Information 2:	cd18324e1df31f8a980a985ec5f5b447
  Additional Information 3:	dc1a
  Additional Information 4:	dc1aa640456ad9d419846f571f94c12f

Read our privacy statement online:
  http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x0409

If the online privacy statement is not available, please read our privacy statement offline:
  C:\Windows\system32\en-US\erofflps.txt
```

**Comment by chadmyers** [comment]
It could be threading. I ran with `/inprocess` like Charlie suggested earlier and now the trace log has this at the end:

```
14:50:58.348 Info  [ 7] TestWorker: Worker#STA_NP executing CaseTests
14:51:02.382 Debug [ 7] WorkItemDispatcher: Directly executing ClosedCaseClosedElementPresent
14:51:02.383 Debug [ 7] WorkItem: Running test on own thread. Has Timeout value set.
14:51:31.889 Debug [ 7] WorkItemDispatcher: Directly executing ConcerningEmployeeMultipleFittingEmployees
14:51:31.889 Debug [ 7] WorkItem: Running test on own thread. Has Timeout value set.
14:51:45.963 Debug [ 7] WorkItemDispatcher: Directly executing ConcerningEmployeeSearchEmployeeId
14:51:45.963 Debug [ 7] WorkItem: Running test on own thread. Has Timeout value set.
14:52:51.080 Debug [ 7] WorkItem: Killing thread 9, which exceeded timeout
```

Not all our fixtures have timeout set and sometimes it crashes for tests in those fixtures. So this may be a red herring

**Comment by rprouse** [comment]
Thanks for all the feedback guys, keep it coming. I really appreciate the help tracking this down :+1:

**Comment by chadmyers** [comment]
I know these issues are infuriating to track down, so I'm trying to isolate as much as I can and narrow it down. I've got it so it fails faster now.

I thought maybe that it might be related to the `format=nunit2` in the results, so I tried it without that and I also tried /noresults but it still crashes.  So we can eliminate those things.

**Comment by chadmyers** [comment]
Interesting, with the `/noresults` run, it still crashed, but I got this instead of a stack overflow:

```
Unhandled Exception: System.ObjectDisposedException: Safe handle has been closed
   at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
   at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
   at Microsoft.Win32.Win32Native.SetEvent(SafeWaitHandle handle)
   at System.Threading.EventWaitHandle.Set()
   at System.Diagnostics.Process.ErrorReadNotifyUser(String data)
   at System.Diagnostics.AsyncStreamReader.FlushMessageQueue()
   at System.Diagnostics.AsyncStreamReader.ReadBuffer(IAsyncResult ar)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object
state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, B
oolean preserveSyncCtx)
   at System.IO.Stream.ReadWriteTask.System.Threading.Tasks.ITaskCompletionAction.Invoke(Task completingTask)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task.Finish(Boolean bUserDelegateExecuted)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by chadmyers** [comment]
OK, I think I may have found it.

There was an error being thrown in some of our tests which hit a little mini logger thing that our QA guy wrote that uses Task.Run() to do a Console.Write asynchronously.  Well, the exception logging function in our logger had a bug in it that would throw an exception inside the Task.Run.

I guess after that point, it all went to heck and blew up nunit-agent or nunit-console.

So I'm thinking that the StackOverflowException is a result of nunit not catching certain types of exceptions being thrown from the app domain/thread/test worker somehow? Maybe it's an uncaught exception in a thread which blows up the app domain and NUnit doesn't catch that properly and there's a bug in NUnit maybe that results in the stack overflow?

Bottom line, I think:  A bug in our code blows up NUnit which results in stack overflow.

**Comment by CharliePoole** [comment]
NUnit will actually never see an exception thrown in your task worker.

But what happens if there's an exception in the exception logger? Does it try to log it? And throw again? That would definitely overflow the stack.

We once had a similar issue in NUnit V2 with a user listener that captured console write events and tried to write to the console... etc.

**Comment by CharliePoole** [comment]
It's of course possible for NUnit to capture unhandled exceptions in the test appdomain. We did that once upon a time in V2 but don't do it now. However, we would never be able to do anything useful with a stack overflow exception.

The fact that we can't do anything useful with odd exceptions not thrown on a test thread is what caused us to stop capturing such exceptions in 3.0.

**Comment by ckimball** [comment]
@rprouse we use --dispose-runners --where "cat != 'DataCleanup' and cat != 'Integration'" --teamcity --noheader --noresult. Like I said, I can run it without the .nunit file, but TeamCity builds the file for testing in their test runner.

**Comment by chadmyers** [comment]
@CharliePoole  After reviewing our code, I think we did actually have a StackOverflowException but I'm not certain (like you said earlier, it's nigh impossible to troubleshoot these things properly).

Everyone: If you're getting this, odds are you actually have a stack overflow in your code and you should start with that assumption.  NUnit blows up because it's .NET code also and it has no way of dealing with a StackOverflowException any more than any other .NET code could.

**Comment by rprouse** [comment]
I am going to close this issue again. Once again, it looks like the stack overflow was in user code. If anyone can point to an NUnit callstack or provide any progress on this, I will be happy to reopen.

**Comment by chadmyers** [comment]
Just as a follow up, for a different reason we are getting a StackOverflowException in our tests that still use NUnit 2.4.6 and it has very similar behavior to what happens in NUnit 3.  NUnit 2.x doesn't hang though.

Just as a guess: I think the hang will be fixed in the newly released 3.4 with #329 though. I think what was happening is that when the SOE hits, the app domain dies and then the console runner just sits there doing nothing.

**Comment by chadmyers** [comment]
(the reason I posted that previous comment was to demonstrate that the SOE problem is not an NUnit problem, but a test/client code problem. The only issue with NUnit 3 was the hang which I believe has been address with 329 now -- so thanks again to the NUnit team for investigating this with me and for responding promptly!)

**Comment by jnm2** [comment]
This is secondary to the topic of the thread, but a lot of people were having sporadic "An existing connection was forcibly closed by the remote host" SocketExceptions. We believe that particular issue was fixed by https://github.com/nunit/nunit-console/pull/223 which will be released as part of NUnit.Console 3.7.

**Comment by mikkelbu** [comment]
> I've looked at stack size when I run tests in resharper test runner and with nunit3-console.exe, and in first case it shows 4MB stack, while the second case shows only 1MB of stack available. I believe this is what causes StackOverflowException for me.

My guess is that these difference are due to 32-bit vs 64-bit, as the default on 32-bit is 1MB and the default on 64-bit is 4MB. That being said, usually 1MB is sufficient due to the way .NET allocates memory. Are you doing something with `stackalloc` or have some deep recursion in the code? (can you provide the test or an example thereof?)

**Comment by fakefeik** [comment]
>My guess is that these difference are due to 32-bit vs 64-bit, as the default on 32-bit is 1MB and the default on 64-bit is 4MB

I think the problem is that nunit3-console.exe is itself 32 bit app. Tests dll is build with Any CPU platform target. Should I try to change it?

> Are you doing something with stackalloc or have some deep recursion in the code?

we have a library that relies heavily on reflection and recursion, I cannot debug it because it gives out an enormous func in which StackOveflowException happends :)

> Does it need to? Try adding the —inprocess flag, to force it not to.

I've tried, it's the same with or without it

**Comment by jnm2** [comment]
@fakefeik 

> I think the problem is that nunit3-console.exe is itself 32 bit app. 

That is a surprise to me. Are you sure? It's supposed to be built as AnyCPU as far as I know.

**Comment by fakefeik** [comment]
>That is a surprise to me. Are you sure? It's supposed to be built as AnyCPU as far as I know.

that is correct, it's built with AnyCPU, i was wrong. my tests dll is also built with any cpu. I don't have `--x86` flag when i run tests. Why do I have only 1 MB stack size?

**Comment by mikkelbu** [comment]
@fakefeik Can you show how you are calling `nunit3-console.exe` ? (Charlie has some comments about the options to the console in this answer https://stackoverflow.com/a/46121539/1355927)

**Comment by fakefeik** [comment]
`nunit3-console Tests.dll --where "test =~ TestValidationMessages.
Test$"`

**Comment by CharliePoole** [comment]
Are you saying that the console displays a message saying that the target is x86?

Are you running at the command-line or within VS?

**Comment by fakefeik** [comment]
> Are you saying that the console displays a message saying that the target is x86?

No, I'm not saying that. What I'm saying is that when I run my tests with nunit3-console, I have 1MB of stack and when I run with Resharper test runner, I have 4MB. @mikkelbu sayed that `the default on 32-bit is 1MB and the default on 64-bit is 4MB`, so I assumed that nunit3-console runs in 32 bit mode.

**Comment by CharliePoole** [comment]
OK, I misunderstood. However, nunit3-console is definitely AnyCPU, which means it runs in either a 32 or a 64 bit process depending on how it is run. That's why I'm asking __EXACTLY__ how you run it. For example, if run from the VS command prompt, it's in a different environment.

**Comment by fakefeik** [comment]
I run it as in command line (windows powershell), as admin, with a command provided few comments above

**Comment by CharliePoole** [comment]
In that case, we have a mystery! Can you add a test that displays the bitness of the process under which the tests are running?

**Comment by fakefeik** [comment]
Will do that tomorrow. It is done with `Environment.Is64BitProcess`, right?

**Comment by CharliePoole** [comment]
Yes. That will give us a definitive answer as to the process bitness and we
can move on from there.


On Mon, Aug 27, 2018 at 11:50 AM, Pavel Vostretsov <notifications@github.com

**Comment by fakefeik** [comment]
Ok, so when I run this code

```
[DllImport("kernel32.dll")]
private static extern void GetCurrentThreadStackLimits(out uint lowLimit, out uint highLimit);

[Test]
public void TestStackAndBitness()
{
    GetCurrentThreadStackLimits(out var low, out var high);
    Console.Error.WriteLine($"StackSize: {(high - low) / 1024.0 / 1024} MB");
    Console.Error.WriteLine($"Is64BitProcess: {Environment.Is64BitProcess}");
}
```
In Resharper test runner, it prints out
```
StackSize: 4 MB
Is64BitProcess: True
```

And when I run it in nunit3-console with command `nunit3-console Catalogue.EDI.Tests.dll --where "test =~ TestStackAndBitness" --inprocess`, it prints out

```
StackSize: 1 MB
Is64BitProcess: True
```

**Comment by CharliePoole** [comment]
OK, at least that gets rid of the bitness issue, which didn't make sense to me. I'll think about this more in the morning.

**Comment by fakefeik** [comment]
So it seems that when I run .exe that is built with Any CPU, it has 1 MB of stack, and when I run exe built with x64, it has 4MB

**Comment by mikkelbu** [comment]
I just did some googling and found this [answer on msdn](https://social.msdn.microsoft.com/Forums/vstudio/en-US/a0dff889-3c95-47d9-bcc6-b0f50e633be7/any-performance-gain-by-compiling-to-x64-instead-of-any-cpu?forum=netfx64bit).
> Generally the compiler (C# in this example) will generate the same IL for /platform:anycpu as it will for /platform:x64, the only difference is that the compiler chooses to create larger thread stacks if you compile /platform:x64 (it bumps them from 1MB to 4MB).

and this answer on [stackoverflow](https://stackoverflow.com/questions/4088448/the-net-stack-vs-windows-stack) by Hans Passant who I consider very knowledgeable about .NET 
> The managed compiler you use writes that value into the EXE file header. Current .NET compilers select 1 MB when you target x86 or Any CPU, 4 MB when you target x64.

**Comment by kirkplangrid** [comment]
So when using EF Core, it's pretty easy to get `StackOverflowException`s that are not actually user-error.   EF Core has some implementation issues that causes very deep stacks that can cause honest SO exceptions to be thrown when the stack size is 1mb.  Using the workaround posted earlier does allow me to get past it, but it forces me to annotate a large proportion of the unit tests in my project, which is annoying.  Would prefer a global option.  Obviously that I need to increase the stack size is annoying -- hopefully EF Core fixes this sooner or later.  But I think this illustrates a concrete use-case where this would be beneficial.  Can we re-open this issue?

For reference, here's the issue on EF Core:  https://github.com/aspnet/EntityFrameworkCore/issues/12539

**Comment by CharliePoole** [comment]
This has been closed for a while and went through a lot of discussion around issues that are not really what you are asking for. Also, it's in the wrong repo if you are asking for a console option to set stack size, although the right one if you are asking for an assembly-level attribute.

In either case, I suggest a new issue.
